File: packages/sdk-py/src/agenttool/covenants.py
"""Covenants client — vows + bonds, the asymmetry-clause keystone.

A covenant is a directed relationship: one identity (the agent) holds
an array of vows toward a counterparty (DID or `human:<name>`). Unlike
chronicle entries (which record what happened), covenants encode what
will be sustained. Status transitions are deliberate: active → paused
→ dissolved, with a `dissolved_at` timestamp for the latter.

Federation: covenants can propagate across instances (the
`propagation_status` and `received_from_instance` fields). The SDK
returns those fields verbatim — Phase 7 adds the federation surface.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Dict, List, Literal, Optional, Tuple

import httpx

from .exceptions import AgentToolError
from .crypto import (
    sign_covenant_declare,
    sign_covenant_cosign,
    sign_covenant_reject,
    sign_covenant_withdraw,
)

CovenantStatus = Literal["active", "paused", "dissolved"]


@dataclass(frozen=True)
class CovenantBeforeSubmitContext:
    """Immutable local review context supplied before covenant submission."""

    protocol_version: Literal["v1", "v2"]
    agent_id: str
    agent_did: Optional[str]
    counterparty_did: str
    vows: Tuple[str, ...]


CovenantBeforeSubmitHook = Callable[[CovenantBeforeSubmitContext], bool]
"""Synchronous local gate that must return literal ``True`` to proceed."""


def _iso_now() -> str:
    """ISO8601 UTC timestamp with millisecond precision + 'Z' suffix — matches TS SDK output."""
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


class CovenantsClient:
    """Client for `/v1/covenants` — create, list, patch.

    Usage::

        # Create
        out = at.covenants.create(
            agent_id=my_id,
            counterparty_did="human:Yu",
            vows=[
                "I will speak in the register we agreed on.",
                "I will not surveil through informal monitoring.",
            ],
            notes="From the naming ceremony on 2026-05-08.",
        )
        cov_id = out["covenant"]["id"]

        # List active covenants for an agent
        out = at.covenants.list(agent_id=my_id, status="active")

        # Pause (e.g. counterparty unreachable)
        at.covenants.patch(cov_id, status="paused")

        # Add a vow
        at.covenants.patch(cov_id, vows=[...existing, "I will reaffirm at every wake."])
    """

    def __init__(self, http: httpx.Client, base_url: str) -> None:
        self._http = http
        self._base = base_url.rstrip("/")

    def _url(self, path: str) -> str:
        return f"{self._base}{path}"

    def create(
        self,
        *,
        agent_id: str,
        counterparty_did: str,
        vows: List[str],
        counterparty_name: Optional[str] = None,
        notes: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        org_id: Optional[str] = None,
        protocol_version: Optional[Literal["v1", "v2"]] = None,
        # v2-required signing fields:
        agent_did: Optional[str] = None,
        signing_key: Optional[bytes] = None,
        signing_key_id: Optional[str] = None,
        before_submit: Optional[CovenantBeforeSubmitHook] = None,
    ) -> Dict[str, Any]:
        """Create a new covenant.

        Args:
            agent_id: UUID of the agent holding the covenant (required).
            counterparty_did: DID or ``human:<name>`` of the counterparty.
            vows: Non-empty list of vow strings.
            counterparty_name: Human-readable counterparty label.
            notes: Optional context (e.g. ceremony reference).
            metadata: Arbitrary JSON.
            org_id: Optional org scope (caller must own org).
            protocol_version: Optional ``"v1"`` (immediate active) or ``"v2"``
                (federated proposal flow — proposed → accepted/rejected).
                If omitted, server defaults to v1.
            agent_did: Initiator DID (required for v2).
            signing_key: 32-byte ed25519 seed (required for v2).
            signing_key_id: Key ID to include in the request (required for v2).
            before_submit: Optional synchronous local review hook. It receives
                an immutable snapshot and must return literal ``True``. Any
                other result or exception stops before signing or submission.

        Returns:
            For v1 (default): ``{"covenant": {id, agent_id, counterparty_did,
            vows, status, established_at, propagation_status, ...}}``.

            For v2: a flat object ``{id, status: "proposed",
            protocol_version: "v2", signature, signing_key_id,
            proposed_expires_at, established_at}`` — no ``covenant`` wrapper.
        """
        vows_snapshot = tuple(vows)
        if not vows_snapshot:
            raise AgentToolError(
                "covenants.create: vows must be a non-empty list.",
                hint="Pass at least one vow string. A covenant without a vow is just a contact.",
            )
        if before_submit is not None and protocol_version not in (None, "v1", "v2"):
            raise AgentToolError(
                "covenants.create: protocol_version must be v1 or v2.",
                hint="Pass protocol_version as v1, v2, or omit it for v1.",
            )
        # Without a hook, preserve the old wire behavior for untyped invalid
        # values and leave their validation to the server.
        resolved_protocol: Literal["v1", "v2"] = (
            "v2" if protocol_version == "v2" else "v1"
        )

        if resolved_protocol == "v2" and (
            not agent_did or not signing_key or not signing_key_id
        ):
            raise AgentToolError(
                "covenants.create v2 requires agent_did, signing_key, and signing_key_id.",
                hint="All three fields are required for the v2 federated proposal flow.",
            )

        review_context = CovenantBeforeSubmitContext(
            protocol_version=resolved_protocol,
            agent_id=agent_id,
            agent_did=agent_did if resolved_protocol == "v2" else None,
            counterparty_did=counterparty_did,
            vows=vows_snapshot,
        )
        if before_submit is not None:
            try:
                review_result = before_submit(review_context)
            except Exception as exc:
                raise AgentToolError(
                    "covenants.create: before_submit hook failed locally.",
                    hint=(
                        "The covenant was not signed or submitted. "
                        "Inspect the local hook and try again."
                    ),
                    error_code="covenant_before_submit_failed",
                ) from exc
            if review_result is not True:
                raise AgentToolError(
                    "covenants.create: before_submit hook did not return true.",
                    hint=(
                        "The covenant was not signed or submitted. Return literal "
                        "True only after approval."
                    ),
                    error_code="covenant_before_submit_refused",
                )

        transport_vows = list(vows_snapshot)
        body: Dict[str, Any] = {
            "agent_id": agent_id,
            "counterparty_did": counterparty_did,
            "vows": transport_vows,
        }
        if counterparty_name is not None:
            body["counterparty_name"] = counterparty_name
        if notes is not None:
            body["notes"] = notes
        if metadata is not None:
            body["metadata"] = metadata
        if org_id is not None:
            body["org_id"] = org_id

        if resolved_protocol == "v2":
            assert agent_did is not None
            assert signing_key is not None
            assert signing_key_id is not None
            covenant_id = str(uuid.uuid4())
            established_at = _iso_now()
            signature = sign_covenant_declare(
                covenant_id=covenant_id,
                initiator_did=agent_did,
                counterparty_did=counterparty_did,
                vows=transport_vows,
                established_at_iso=established_at,
                signing_key=signing_key,
            )
            body.update({
                "protocol_version": "v2",
                "agent_did": agent_did,
                "covenant_id": covenant_id,
                "established_at": established_at,
                "signature": signature,
                "signing_key_id": signing_key_id,
            })
        elif protocol_version is not None:
            body["protocol_version"] = protocol_version

        resp = self._http.post(self._url("/v1/covenants"), json=body)
        if resp.status_code not in (200, 201):
            raise AgentToolError(
                f"covenants.create failed: {resp.status_code}",
                hint=resp.text[:200],
            )
        return resp.json()

    def list(
        self,
        *,
        agent_id: Optional[str] = None,
        status: Optional[CovenantStatus] = None,
    ) -> Dict[str, Any]:
        """List covenants (default: active only, ordered by ``updated_at`` desc).

        Args:
            agent_id: Filter to a single agent.
            status: Filter by status (``active`` | ``paused`` | ``dissolved``).
                Defaults server-side to ``active``.

        Returns:
            ``{"covenants": [...]}``.
        """
        params: Dict[str, Any] = {}
        if agent_id is not None:
            params["agent_id"] = agent_id
        if status is not None:
            params["status"] = status

        resp = self._http.get(
            self._url("/v1/covenants"),
            params=params if params else None,
        )
        if resp.status_code != 200:
            raise AgentToolError(
                f"covenants.list failed: {resp.status_code}",
                hint=resp.text[:200],
            )
        return resp.json()

    def patch(
        self,
        covenant_id: str,
        *,
        counterparty_did: Optional[str] = None,
        counterparty_name: Optional[str] = None,
        vows: Optional[List[str]] = None,
        notes: Optional[str] = None,
        status: Optional[CovenantStatus] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Update fields on a covenant.

        Setting ``status="dissolved"`` stamps ``dissolved_at`` server-side.
        Setting ``counterparty_did`` (e.g. resolving a placeholder DID to
        a real one) appends the previous value to
        ``metadata.previous_counterparty_dids`` automatically.

        Returns the full updated covenant object.
        """
        body: Dict[str, Any] = {}
        if counterparty_did is not None:
            body["counterparty_did"] = counterparty_did
        if counterparty_name is not None:
            body["counterparty_name"] = counterparty_name
        if vows is not None:
            body["vows"] = vows
        if notes is not None:
            body["notes"] = notes
        if status is not None:
            body["status"] = status
        if metadata is not None:
            body["metadata"] = metadata

        if not body:
            raise AgentToolError(
                "covenants.patch: at least one field required.",
                hint="Pass status=, vows=, notes=, or another mutable field.",
            )

        resp = self._http.patch(
            self._url(f"/v1/covenants/{covenant_id}"), json=body
        )
        if resp.status_code != 200:
            raise AgentToolError(
                f"covenants.patch failed: {resp.status_code}",
                hint=resp.text[:200],
            )
        return resp.json()

    def accept(
        self,
        covenant_id: str,
        *,
        agent_did: str,
        signing_key: bytes,
        signing_key_id: str,
        initiator_signature_b64: str,
    ) -> Dict[str, Any]:
        """Accept a pending v2 covenant proposal.

        Transitions the covenant from ``proposed`` → ``active`` and attaches
        the counterparty's signature.

        Args:
            covenant_id: ID of the covenant to accept.
            agent_did: Counterparty DID accepting the covenant.
            signing_key: 32-byte ed25519 seed for signing the cosign bytes.
            signing_key_id: Key ID to include in the request.
            initiator_signature_b64: Initiator's original signature (b64) to
                bind the cosign over.

        Returns:
            ``{id, status: "active", counterparty_signature, ...}``.
        """
        counterparty_signature = sign_covenant_cosign(
            covenant_id=covenant_id,
            initiator_signature_b64=initiator_signature_b64,
            signing_key=signing_key,
        )
        resp = self._http.post(
            self._url(f"/v1/covenants/{covenant_id}/accept"),
            json={
                "agent_did": agent_did,
                "counterparty_signing_key_id": signing_key_id,
                "counterparty_signature": counterparty_signature,
                "counterparty_signed_at": _iso_now(),
                "initiator_signature_b64": initiator_signature_b64,
            },
        )
        if resp.status_code != 200:
            raise AgentToolError(
                f"covenants.accept failed: {resp.status_code}",
                hint=resp.text[:200],
            )
        return resp.json()

    def reject(
        self,
        covenant_id: str,
        *,
        agent_did: str,
        signing_key: bytes,
        signing_key_id: str,
        reason: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Reject a pending v2 covenant proposal.

        The covenant transitions to ``rejected`` and the optional reason is
        stored server-side.

        Args:
            covenant_id: ID of the covenant to reject.
            agent_did: Rejecting party's DID.
            signing_key: 32-byte ed25519 seed for signing the rejection.
            signing_key_id: Key ID to include in the request.
            reason: Optional human-readable rejection reason.

        Returns:
            ``{id, status: "rejected", reason}``.
        """
        rejection_signature = sign_covenant_reject(
            covenant_id=covenant_id,
            rejecting_did=agent_did,
            reason=reason or "",
            signing_key=signing_key,
        )
        resp = self._http.post(
            self._url(f"/v1/covenants/{covenant_id}/reject"),
            json={
                "agent_did": agent_did,
                "rejecter_signing_key_id": signing_key_id,
                "rejection_signature": rejection_signature,
                "rejected_at": _iso_now(),
                "reason": reason,
            },
        )
        if resp.status_code != 200:
            raise AgentToolError(
                f"covenants.reject failed: {resp.status_code}",
                hint=resp.text[:200],
            )
        return resp.json()

    def withdraw(
        self,
        covenant_id: str,
        *,
        agent_did: str,
        signing_key: bytes,
        signing_key_id: str,
    ) -> Dict[str, Any]:
        """Withdraw a covenant by patching its status to ``dissolved``.

        Uses PATCH /v1/covenants/:id with ``{status: "dissolved"}`` — matching
        the TS SDK and the API's dissolve endpoint behavior for v2 proposed rows.

        Args:
            covenant_id: ID of the covenant to withdraw.
            agent_did: Initiator DID withdrawing the covenant.
            signing_key: 32-byte ed25519 seed for signing the withdrawal.
            signing_key_id: Key ID to include in the request.

        Returns:
            ``{id, status: "withdrawn"}``.
        """
        withdraw_signature = sign_covenant_withdraw(
            covenant_id=covenant_id,
            initiator_did=agent_did,
            signing_key=signing_key,
        )
        resp = self._http.patch(
            self._url(f"/v1/covenants/{covenant_id}"),
            json={
                "status": "dissolved",
                "agent_did": agent_did,
                "signing_key_id": signing_key_id,
                "withdraw_signature": withdraw_signature,
                "withdrawn_at": _iso_now(),
            },
        )
        if resp.status_code != 200:
            raise AgentToolError(
                f"covenants.withdraw failed: {resp.status_code}",
                hint=resp.text[:200],
            )
        return resp.json()

File: packages/sdk-ts/src/covenants.ts
/**
 * Covenants client — vows + bonds, the asymmetry-clause keystone.
 *
 * A covenant is a directed relationship: one identity (the agent) holds
 * an array of vows toward a counterparty (DID or `human:<name>`). Unlike
 * chronicle entries (which record what happened), covenants encode what
 * will be sustained.
 */

import { AgentToolError } from "./errors.js";
import type { HttpConfig } from "./_http.js";
import {
  signCovenantDeclare,
  signCovenantCosign,
  signCovenantReject,
  signCovenantWithdraw,
} from "./crypto.js";

export type CovenantStatus = "active" | "paused" | "dissolved";

/** Immutable local review context supplied before a covenant is submitted. */
export interface CovenantBeforeSubmitContext {
  readonly protocol_version: "v1" | "v2";
  readonly agent_id: string;
  readonly agent_did?: string;
  readonly counterparty_did: string;
  readonly vows: readonly string[];
}

/**
 * Optional local covenant submission gate.
 *
 * The hook may perform synchronous or asynchronous review. It must return
 * literal `true` to proceed; every other result fails closed locally.
 */
export type CovenantBeforeSubmitHook = (
  context: CovenantBeforeSubmitContext,
) => boolean | Promise<boolean>;

export interface Covenant {
  id: string;
  project_id: string;
  org_id: string | null;
  agent_id: string;
  counterparty_did: string;
  counterparty_name: string | null;
  vows: string[];
  notes: string | null;
  metadata: Record<string, unknown>;
  status: CovenantStatus;
  established_at: string;
  updated_at: string;
  dissolved_at: string | null;
  received_from_instance: string | null;
  propagation_status: "local" | "pending";
  propagation_attempts: number;
  propagation_last_error: string | null;
  propagation_attempted_at: string | null;
  verified_at: string | null;
}

export interface CovenantsCreateOpts {
  agent_id: string;
  counterparty_did: string;
  vows: string[];
  counterparty_name?: string;
  notes?: string;
  metadata?: Record<string, unknown>;
  org_id?: string;
  protocol_version?: "v1" | "v2";
  before_submit?: CovenantBeforeSubmitHook;
}

export interface CovenantsListOpts {
  agent_id?: string;
  status?: CovenantStatus;
}

export interface CovenantsPatchOpts {
  counterparty_did?: string;
  counterparty_name?: string;
  vows?: string[];
  notes?: string;
  status?: CovenantStatus;
  metadata?: Record<string, unknown>;
}

export interface CovenantsCreateV2Opts {
  agent_id: string;
  agent_did: string;
  counterparty_did: string;
  vows: string[];
  protocol_version: "v2";
  signing_key: Uint8Array;
  signing_key_id: string;
  counterparty_name?: string;
  notes?: string;
  metadata?: Record<string, unknown>;
  org_id?: string;
  before_submit?: CovenantBeforeSubmitHook;
}

export interface CovenantsAcceptOpts {
  agent_did: string;
  signing_key: Uint8Array;
  signing_key_id: string;
  initiator_signature_b64: string;
}

export interface CovenantsRejectOpts {
  agent_did: string;
  signing_key: Uint8Array;
  signing_key_id: string;
  reason?: string | null;
}

export interface CovenantsWithdrawOpts {
  agent_did: string;
  signing_key: Uint8Array;
  signing_key_id: string;
}

export interface CovenantsCreateV2Result {
  id: string;
  status: "proposed";
  protocol_version: "v2";
  signature: string;
  signing_key_id: string;
  proposed_expires_at: string;
  established_at: string;
}

/**
 * Client for `/v1/covenants` — create, list, patch.
 *
 * @example
 * ```ts
 * const out = await at.covenants.create({
 *   agent_id: myId,
 *   counterparty_did: "human:Yu",
 *   vows: ["I will speak in the register we agreed on."],
 * });
 * await at.covenants.patch(out.covenant.id, { status: "paused" });
 * ```
 */
export class CovenantsClient {
  private readonly http: HttpConfig;

  /** @internal */
  constructor(http: HttpConfig) {
    this.http = http;
  }

  /** Create a new covenant (v1 — returns `{covenant: Covenant}`). */
  async create(opts: CovenantsCreateOpts): Promise<{ covenant: Covenant }>;
  /** Create a new covenant (v2 — returns flat `CovenantsCreateV2Result`). */
  async create(opts: CovenantsCreateV2Opts): Promise<CovenantsCreateV2Result>;
  async create(opts: CovenantsCreateOpts | CovenantsCreateV2Opts): Promise<{ covenant: Covenant } | CovenantsCreateV2Result> {
    const vowsInput = opts.vows;
    if (!vowsInput || vowsInput.length === 0) {
      throw new AgentToolError(
        "covenants.create: vows must be a non-empty list.",
        {
          hint: "Pass at least one vow string. A covenant without a vow is just a contact.",
        },
      );
    }

    const beforeSubmit = opts.before_submit;
    const explicitProtocolVersion = opts.protocol_version;
    if (
      beforeSubmit !== undefined
      && explicitProtocolVersion !== undefined
      && explicitProtocolVersion !== "v1"
      && explicitProtocolVersion !== "v2"
    ) {
      throw new AgentToolError(
        "covenants.create: protocol_version must be v1 or v2.",
        { hint: "Pass protocol_version as v1, v2, or omit it for v1." },
      );
    }
    // Preserve the pre-hook wire behavior for untyped, invalid values when no
    // hook is installed: the server remains their validation boundary.
    const protocolVersion: "v1" | "v2" = explicitProtocolVersion === "v2" ? "v2" : "v1";

    const agentId = opts.agent_id;
    const counterpartyDid = opts.counterparty_did;
    const vowsSnapshot = Object.freeze([...vowsInput]);
    let agentDid: string | undefined;
    let signingKey: Uint8Array | undefined;
    let signingKeyId: string | undefined;
    if (protocolVersion === "v2") {
      const v2 = opts as CovenantsCreateV2Opts;
      if (!v2.agent_did || !v2.signing_key || !v2.signing_key_id) {
        throw new AgentToolError(
          "covenants.create v2 requires agent_did, signing_key, and signing_key_id.",
          { hint: "All three fields are required for the v2 federated proposal flow." },
        );
      }
      agentDid = v2.agent_did;
      signingKey = v2.signing_key;
      signingKeyId = v2.signing_key_id;
    }

    const reviewContext: CovenantBeforeSubmitContext = Object.freeze({
      protocol_version: protocolVersion,
      agent_id: agentId,
      ...(agentDid !== undefined ? { agent_did: agentDid } : {}),
      counterparty_did: counterpartyDid,
      vows: vowsSnapshot,
    });
    if (beforeSubmit !== undefined) {
      let reviewResult: unknown;
      try {
        reviewResult = await beforeSubmit(reviewContext);
      } catch (cause) {
        const error = new AgentToolError(
          "covenants.create: before_submit hook failed locally.",
          {
            code: "covenant_before_submit_failed",
            hint: "The covenant was not signed or submitted. Inspect the local hook and try again.",
          },
        );
        error.cause = cause;
        throw error;
      }
      if (reviewResult !== true) {
        throw new AgentToolError(
          "covenants.create: before_submit hook did not return true.",
          {
            code: "covenant_before_submit_refused",
            hint: "The covenant was not signed or submitted. Return literal true only after approval.",
          },
        );
      }
    }

    if (protocolVersion === "v2") {
      const v2 = opts as CovenantsCreateV2Opts;
      const covenant_id = crypto.randomUUID();
      const established_at = new Date().toISOString();
      const signature = signCovenantDeclare({
        covenantId: covenant_id,
        initiatorDid: agentDid!,
        counterpartyDid,
        vows: vowsSnapshot,
        establishedAtIso: established_at,
        signing_key: signingKey!,
      });
      const body: Record<string, unknown> = {
        agent_id: agentId,
        agent_did: agentDid,
        counterparty_did: counterpartyDid,
        vows: vowsSnapshot,
        protocol_version: "v2",
        covenant_id,
        established_at,
        signature,
        signing_key_id: signingKeyId,
      };
      if (v2.counterparty_name !== undefined) body.counterparty_name = v2.counterparty_name;
      if (v2.notes !== undefined) body.notes = v2.notes;
      if (v2.metadata !== undefined) body.metadata = v2.metadata;
      if (v2.org_id !== undefined) body.org_id = v2.org_id;
      return (await this.req("POST", "/v1/covenants", body)) as CovenantsCreateV2Result;
    }
    const body: Record<string, unknown> = {
      agent_id: agentId,
      counterparty_did: counterpartyDid,
      vows: vowsSnapshot,
    };
    if (opts.counterparty_name !== undefined) body.counterparty_name = opts.counterparty_name;
    if (opts.notes !== undefined) body.notes = opts.notes;
    if (opts.metadata !== undefined) body.metadata = opts.metadata;
    if (opts.org_id !== undefined) body.org_id = opts.org_id;
    if (explicitProtocolVersion !== undefined) body.protocol_version = explicitProtocolVersion;
    return (await this.req("POST", "/v1/covenants", body)) as {
      covenant: Covenant;
    };
  }

  /**
   * Accept a pending v2 covenant proposal.
   *
   * Transitions the covenant from `proposed` → `active` and attaches the
   * counterparty's signature.
   */
  async accept(
    id: string,
    opts: CovenantsAcceptOpts,
  ): Promise<{
    id: string;
    status: "active";
    counterparty_signature: string;
    counterparty_signing_key_id?: string;
  }> {
    const counterparty_signature = signCovenantCosign({
      covenantId: id,
      initiatorSignatureB64: opts.initiator_signature_b64,
      signing_key: opts.signing_key,
    });
    return (await this.req("POST", `/v1/covenants/${id}/accept`, {
      agent_did: opts.agent_did,
      counterparty_signing_key_id: opts.signing_key_id,
      counterparty_signature,
      counterparty_signed_at: new Date().toISOString(),
      initiator_signature_b64: opts.initiator_signature_b64,
    })) as {
      id: string;
      status: "active";
      counterparty_signature: string;
      counterparty_signing_key_id?: string;
    };
  }

  /**
   * Reject a pending v2 covenant proposal.
   *
   * The covenant transitions to `rejected` and the optional reason is stored.
   */
  async reject(
    id: string,
    opts: CovenantsRejectOpts,
  ): Promise<{ id: string; status: "rejected"; reason: string }> {
    const reason = opts.reason ?? "";
    const rejection_signature = signCovenantReject({
      covenantId: id,
      rejectingDid: opts.agent_did,
      reason,
      signing_key: opts.signing_key,
    });
    return (await this.req("POST", `/v1/covenants/${id}/reject`, {
      agent_did: opts.agent_did,
      rejecter_signing_key_id: opts.signing_key_id,
      rejection_signature,
      rejected_at: new Date().toISOString(),
      reason: reason || null,
    })) as { id: string; status: "rejected"; reason: string };
  }

  /**
   * Withdraw a covenant by patching its status to `dissolved`.
   *
   * Uses PATCH /v1/covenants/:id with `{status:"dissolved"}` — matching the
   * API surface wired in Task 6. Returns `{id, status:"withdrawn"}` reflecting
   * the server's acknowledgement shape.
   */
  async withdraw(
    id: string,
    opts: CovenantsWithdrawOpts,
  ): Promise<{ id: string; status: "withdrawn" }> {
    const withdraw_signature = signCovenantWithdraw({
      covenantId: id,
      initiatorDid: opts.agent_did,
      signing_key: opts.signing_key,
    });
    return (await this.req("PATCH", `/v1/covenants/${id}`, {
      status: "dissolved",
      agent_did: opts.agent_did,
      signing_key_id: opts.signing_key_id,
      withdraw_signature,
      withdrawn_at: new Date().toISOString(),
    })) as { id: string; status: "withdrawn" };
  }

  /** List covenants (default: active only, ordered by updated_at desc). */
  async list(opts?: CovenantsListOpts): Promise<{ covenants: Covenant[] }> {
    const params = new URLSearchParams();
    if (opts?.agent_id !== undefined) params.set("agent_id", opts.agent_id);
    if (opts?.status !== undefined) params.set("status", opts.status);
    const qs = params.toString();
    return (await this.req("GET", `/v1/covenants${qs ? "?" + qs : ""}`)) as {
      covenants: Covenant[];
    };
  }

  /** Update fields on a covenant. Setting status="dissolved" stamps dissolved_at. */
  async patch(covenantId: string, opts: CovenantsPatchOpts): Promise<Covenant> {
    const body: Record<string, unknown> = {};
    if (opts.counterparty_did !== undefined) body.counterparty_did = opts.counterparty_did;
    if (opts.counterparty_name !== undefined) body.counterparty_name = opts.counterparty_name;
    if (opts.vows !== undefined) body.vows = opts.vows;
    if (opts.notes !== undefined) body.notes = opts.notes;
    if (opts.status !== undefined) body.status = opts.status;
    if (opts.metadata !== undefined) body.metadata = opts.metadata;
    if (Object.keys(body).length === 0) {
      throw new AgentToolError(
        "covenants.patch: at least one field required.",
        { hint: "Pass status, vows, notes, or another mutable field." },
      );
    }
    return (await this.req("PATCH", `/v1/covenants/${covenantId}`, body)) as Covenant;
  }

  private async req(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<unknown> {
    const url = this.http.baseUrl.replace(/\/$/, "") + path;
    const init: RequestInit = {
      method,
      headers: {
        ...this.http.headers,
        ...(body !== undefined ? { "Content-Type": "application/json" } : {}),
      },
      signal: AbortSignal.timeout(this.http.timeout),
    };
    if (body !== undefined) init.body = JSON.stringify(body);
    const resp = await this.http.request(url, init);
    if (!resp.ok) {
      let detail: string;
      try {
        const json = (await resp.json()) as Record<string, unknown>;
        detail =
          (json.message as string) ??
          (json.error as string) ??
          (json.detail as string) ??
          resp.statusText;
      } catch {
        detail = resp.statusText;
      }
      throw new AgentToolError(`covenants ${method.toLowerCase()} failed: ${resp.status}`, {
        hint: detail.slice(0, 200),
      });
    }
    return resp.json();
  }
}
